### ISSUE TITLE
BERT4Rec: Sequential Recommendation with Bidirectional Encoder Representations from Transformer

### ISSUE THREAD

**Issue by Yphy** [Yphy/Paper-Review#3]
[BERT4Rec: Sequential Recommendation with Bidirectional Encoder Representations from Transformer](https://arxiv.org/abs/1904.06690)

## Introduction
대부분의 추천모델은 interaction histories에 기반한 CF 모델을 사용한다. 그 중에서도 유저와 아이템의 잠재 벡터 연산을 통해 선호를 예측하는 Matrix Factorization같은 방식이 가장 인기가 있었다. 다른 한 가지 방법은 아이템 기반의 neighborhood 방법이다. 이는 사전에 계산된 아이템들간의 유사도를 활용해 유저의 인터렉션에 기반해 유사한 다른 아이템을 예측하는 방식이다. 

추천 모델 중 sequential neural network는 가령 쇼핑몰 유저의 구매목록 등 sequential한 정보를 이용하여, 그 다음 시점의 유저 interaction에 대해서 예측하는 모델이다. 기존의 GRURec, SASRec 등의 sequential 모델은 left to right로 unidirectional한 방식으로 학습을 했는데 Bidirectional Encoder을 사용하는 BERT를 이용하여 성능을 개선하였다.

## Problem statement

* Su는 user의 interaction sequence 이며  v_t(u)는  user u가 t 시점에 interaction한 아이템 v이다.

* n_u는 user u 의 interaction sequence , 즉 총 interaction length이며
  결과적으로 알고 싶은 것은 다음 시점의 user interaction 인  v_nu+1이다.

## BERT4Rec model architecture

<img width="949" alt="스크린샷 2021-04-08 오후 4 03 50" src="https://user-images.githubusercontent.com/47969237/114198484-58d9b300-998e-11eb-97f5-3ed545493290.png">

-   (a)의 Transformer Layer를 이용하며 Figure1의 아키텍쳐를 봤을 때, (d)의 RNN 처럼 step by step의 형식으로 forward stap의 학습이 진행되는 것이 아니라 parallel하게(layer가 병렬적으로 학습) previous layer 모든 포지션에 대한 정보를 representation 하는 식으로 학습이 되며 네트워크의 핵심이라고 할 수 있다.
-   또한 논문에서는 ***Cloze task***라고 input item을 masking 한다. BERT에서 하는 것과 마찬가지로 학습 시에는 input에 중간중간 랜덤하게 masking 을 하여 학습하고 inference 시에는 최종적으로 예측하고자 하는 마지막 시점의 item을 알기 위해 (b)처럼 Pt만 마스킹을 한다. Train할 때도 똑같이 마지막만 마스킹한 케이스를 학습시킨다고 한다.

## Transformer Layer

* Transformer layer는 Multi-Head Self-Attention sub-layer 와 Position-wise Feed-Forward Network 로 구성된다.
* self attention layer의 output은 feed forward network를 통과하며 position wise- feed forward network 에 적용되는 기법은 크게 세가지이다.
	* dropout
	* layer normalization
	* residual connection
* 또한 활성함수로는 BERT 따라 Gaussian Error Linear Unit(GELU) 를 사용했다. ReLU와 다르게 음수에서도 미분가능하니 약간의 gradient를 얻으려고 한 것이 아닐까

## Positional Embedding
* 위치정보를 주기 위해서 positional Embeddings 수행.
* 아이템 임베딩(vi)과  포지션 임베딩 pi를 더해준다.   $h i^0 = v i + p i$

## Model Learning
* final output representation이 target item 의 정보를 반영하고 있으면 안되기 때문에 t-1 samples를 만드는데 전체 케이스를 생성하는 것은 비효율적이기 때문에 앞에서 설명한 Cloze task를 적용한다. 
	* <img width="491" alt="스크린샷 2021-04-09 오후 8 53 58" src="https://user-images.githubusercontent.com/47969237/114198571-7444be00-998e-11eb-8594-b0dcb375e120.png">

	* masking 비율은 파라미터로 설정
* Loss를 계산할 때 negative sampling 을 적용한다. Word2vec 과 마찬가지로, interaction이 없던 아이템 중에 몇개를 넣어서 loss를 크게주는 방식으로 학습시키는듯.

## Experiment

- Dataset으로는 Steam , Beauty, ML-1m , ML -20m 사용했는데 dataset의 특징에 따른 파라미터 비교값을 잘 비교해봐야 할 것 같다는 생각.

- <img width="1035" alt="스크린샷 2021-04-09 오후 11 41 31" src="https://user-images.githubusercontent.com/47969237/114199084-f339f680-998e-11eb-8fde-367765d9ffbf.png">
 Steam ,Beauty는 user, item 의 수가 커 matrix size는 큰 반면, action이 상대적으로 무비렌즈보다 적어   sparse한 matrix를 예상할 수 있음 즉, low density. 따라서 hidden dimentionality 을 계속해서 키웠을 때 64 이후 overfitting 되는 결과가 보인다. 

- <img width="698" alt="스크린샷 2021-04-09 오후 11 40 44" src="https://user-images.githubusercontent.com/47969237/114199077-f208c980-998e-11eb-9890-0c03d82ab49c.png">
masking 파라미터인 proportion ρ 에 따른 결과. Dataset의 sequence length와 밀접한 영향이 있었는데, steam, beauty가 movielens보다는 짧다고한다 (플레이한 게임수보다는 영화시청이 상대적으로 하기 쉬우니 그런듯). sequence가 짧은 데이터셋에서는 ρ =  0.6일 때, 짧은 경우는 ρ = 0.2일 때, 성능이 가장 좋았음. sequence가 긴데 mask까지 많다면 예측을 너무많이해야 하니 그렇다고 한다.. 

## Conclusion
* Sequential Recommender 계열 논문은 처음 읽는데 느낌탓인지 추천보다는 거의 대부분이 BERT에 관한 내용인 것 같다. 읽다보면 다시 Transformer를 보러 가게된다.
